### PR TITLE
Pyramid: compute and use switch_shape_at_level

### PR DESCRIPTION
**Pull request by lukasdreyer**
**_Describe your changes here:_**

This PR introduces a new field 'switch_shape_at_level' for pyramid elements.
Everytime a pyramid is constructed we set this variable. If the shape of the constructed element is a pyramid, it is set to -1. Otherwise (the shape is a tetrahedron) we set it to the level of last ancestor in the shape of a tetrahedron.
This variable is then used in every other algorithm that needs information about ancestors and speeds up the computation significantly. First tests showed a speed-up around 20% for adapt and balance algorithms, without any changes in the other high-level algorithms.

**_All these boxes must be checked by the reviewers before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is fully understood, bug free, well-documented and well-structured.

- [ ] The author added a BSD statement to `doc/` (or already has one)
- [ ] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)

  If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

- [ ] All tests pass (in various configurations, this should be executed automatically in a github action)
- [ ] New source/header files are properly added to the Makefiles
- [ ] The reviewer executed the new code features at least once and checked the results manually
- [ ] The code is covered in an existing or new test case
- [ ] New tests use the Google Test framework
- [ ] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [ ] The code is well documented
- [ ] All function declarations, structs/classes and their members have a proper doxygen documentation
- [ ] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)
- [ ] Testing of this template: If you feel something is missing from this list, contact the developers

### CODE REVIEW

**Comment by lukasdreyer** [comment]
Draft because https://github.com/DLR-AMR/t8code/pull/347 needs to be merged first

**Comment by lukasdreyer** [comment]
I rebased to main and force pushed because I didn't know that we squash commit PRs.

**Review by jmark — CHANGES_REQUESTED** [review]

File: src/t8_schemes/t8_default/t8_default_pyramid/t8_dpyramid_bits.h [hunk 1]
```diff
@@ -215,7 +206,7 @@ int                 t8_dpyramid_is_inside_root (const t8_dpyramid_t *p);
 int                 t8_dpyramid_is_inside_pyra (const t8_dpyramid_t *p,
```
[the remark this answers is not in the document]
Comment (Davknapp) [reply src/t8_schemes/t8_default/t8_default_pyramid/t8_dpyramid_bits.h:206]: The function `t8_dpyramid_is_inside_root` is needed by the element scheme. I made the others static.

File: src/t8_schemes/t8_default/t8_default_pyramid/t8_dpyramid_bits.h [hunk 2]
```diff
@@ -303,10 +294,9 @@ void                t8_dpyramid_compute_coords (const t8_dpyramid_t *p,
 
 /** Compute the pyramid-parent-type of a tetrahedron
  * \param [in] p        Input pyramid
- * \param [out] parent  The parent whose type has to be computed.
+ * \return              The type of the parent.
  */
-void                t8_dpyramid_tetparent_type (const t8_dpyramid_t *p,
-                                                t8_dpyramid_t *parent);
+int                 t8_dpyramid_tetparent_type (const t8_dpyramid_t *p);
```
[the remark this answers is not in the document]
Comment (Davknapp) [reply src/t8_schemes/t8_default/t8_default_pyramid/t8_dpyramid_bits.h]: It should only be used in `t8_dpyramids_bits.c`

**Review by jmark — APPROVED** [review]

**Pull request merged**